pulsar-package-management/bookkeeper-storage/src/test/java/org/apache/pulsar/packages/management/storage/bookkeeper/bookkeeper/test/ZooKeeperUtil.java

```diff
@@ -16,120 +16,142 @@
+import org.junit.Assert;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.testng.Assert;
-
-import java.io.File;
-import java.io.IOException;
-import java.net.InetSocketAddress;
-import java.util.concurrent.CountDownLatch;
-
-public class ZooKeeperUtil {
-    static final Logger LOG = LoggerFactory.getLogger(org.apache.bookkeeper.test.ZooKeeperUtil.class);
-
-    // ZooKeeper related variables
-    protected static final Integer zooKeeperPort = PortManager.nextFreePort();
-    private final InetSocketAddress zkaddr;
 
+public class ZooKeeperUtil implements ZooKeeperCluster {
+    static final Logger LOG;
+    protected Integer zooKeeperPort = 0;
+    private InetSocketAddress zkaddr;
     protected ZooKeeperServer zks;
-    protected ZooKeeper zkc; // zookeeper client
+    protected ZooKeeper zkc;
     protected NIOServerCnxnFactory serverFactory;
-    protected File ZkTmpDir;
-    private final String connectString;
+    protected File zkTmpDir;
+    private String connectString;
 
     public ZooKeeperUtil() {
-        zkaddr = new InetSocketAddress(zooKeeperPort);
-        connectString = "localhost:" + zooKeeperPort;
+        String loopbackIPAddr = InetAddress.getLoopbackAddress().getHostAddress();
+        this.zkaddr = new InetSocketAddress(loopbackIPAddr, 0);
+        this.connectString = loopbackIPAddr + ":" + this.zooKeeperPort;
     }
 
     public ZooKeeper getZooKeeperClient() {
-        return zkc;
+        return this.zkc;
     }
 
     public String getZooKeeperConnectString() {
-        return connectString;
+        return this.connectString;
+    }
+
+    public String getMetadataServiceUri() {
+        return this.getMetadataServiceUri("/ledgers");
     }
 
-    public void startServer() throws Exception {
-        // create a ZooKeeper server(dataDir, dataLogDir, port)
+    public String getMetadataServiceUri(String zkLedgersRootPath) {
+        return "zk://" + this.connectString + zkLedgersRootPath;
+    }
+
+    public String getMetadataServiceUri(String zkLedgersRootPath, String type) {
+        return "zk+" + type + "://" + this.connectString + zkLedgersRootPath;
+    }
+
+    public void startCluster() throws Exception {
         LOG.debug("Running ZK server");
-        // ServerStats.registerAsConcrete();
         ClientBase.setupTestEnv();
-        ZkTmpDir = File.createTempFile("zookeeper", "test");
-        ZkTmpDir.delete();
-        ZkTmpDir.mkdir();
+        this.zkTmpDir = IOUtils.createTempDir("zookeeper", "test");
+        this.restartCluster();
+        this.createBKEnsemble("/ledgers");
+    }
 
-        zks = new ZooKeeperServer(ZkTmpDir, ZkTmpDir, ZooKeeperServer.DEFAULT_TICK_TIME);
-        serverFactory = new NIOServerCnxnFactory();
-        serverFactory.configure(zkaddr, 100);
-        serverFactory.startup(zks);
+    public void restartCluster() throws Exception {
+        this.zks = new ZooKeeperServer(this.zkTmpDir, this.zkTmpDir, 3000);
```

Comment:
as a followup work we should move to the official ZookKeeper Java API to bootstrap the server from Java (ZooKeeperServerEmbedded) or use Curator Testing Server.
Both of them provide official an tested APIs to run ZooKeeper

Reply:
Good suggestion